pkg/cloudprovider/providers/openstack/openstack_volumes.go

```diff
@@ -317,8 +317,16 @@ func (os *OpenStack) AttachDisk(instanceID, volumeID string) (string, error) {
 		if instanceID == volume.AttachedServerId {
 			glog.V(4).Infof("Disk %s is already attached to instance %s", volumeID, instanceID)
 			return volume.ID, nil
+		} else {
+			// otherwise the disk is in incorrect compute and it needs detach
+			glog.V(2).Infof("Disk %s is attached to a different compute: %s, attempting detach", volumeID, volume.AttachedServerId)
```

Comment:
So this patch unequivocally performs detach from wherever the volume was previously detached in order for attach to succeed? I do not think this is correct approach and might cause race conditions. 

What I would suggest instead is to take approach we took for fixing EBS volumes and raise `volumeutil.NewDanglingError` if volume is attached to a compute node and it shouldn't be attached there - https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/aws/aws.go#L2251

What that will do is, add the volume back into actual state of world of attach/detach controller and if there is no pod that needs the volume, then the volume will be correctly detached.

Reply:
sounds like a plan

Reply:
You are going to update this PR as aforementioned heuristic right?

Reply:
Yep, i will update when i have time which is not now

Reply:
cool. let me place a hold on this PR for now

/hold